Pin 6058 implementazione rotta di creazione istanza

Alepazz opened the pull request:
[PIN-6058](https://pagopa.atlassian.net/browse/PIN-6058)

### Task Objective
Creation of an **eService** instance from an existing **template**. Following this, all associated files are copied from the template's _bucket_ to the **eService**'s _bucket_.

**Note**: Only files of type `DOCUMENTS` are copied, not those of type `INTERFACE`.

---

### Key Points

- Added `.idea` to the .gitignore file (since IntelliJ is used as the development IDE)
- Fixed: the _.env_ file in _eservice_template_readmodel_write_ had incorrect syntax for `ESERVICE_TEMPLATE_TOPIC`
- Added missing fields to the eService model: _EServiceTemplateId_ and _EServiceTemplateVersionId_
- Added a command to start _eservice_template_readmodel_writer_ in the _package.json_
- Created _POST_ APIs to create the new instance: on _catalog_process_ and _bff_

**Note**: When `createEService` is called, some fields are "managed" according to specifications defined along the way. Specifically:

- The _descriptor_ has its `audience` field populated with the template's value
- `dailyCallsPerConsumer` takes its value from the _published_ version of the template; otherwise, it defaults to 1
- `dailyCallsTotal` takes its value from the _published_ version of the template; otherwise, it defaults to 1
- `isConsumerDelegable` and `isClientAccessDelegable` are set to false by default

[PIN-6058]: https://pagopa.atlassian.net/browse/PIN-6058?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Carminepo2 commented on packages/catalog-process/src/services/catalogService.ts:
```diff
@@ -2237,6 +2290,100 @@ export function catalogServiceBuilder(
+            mode: eServiceModeToApiEServiceMode(template.mode),
+            descriptor: {
+              description: publishedVersion.description,
+              audience: [template.audienceDescription],
```
You’re right, the name is definitely misleading.
The audienceDescription refers to the description of who the template is intended for, not the audience itself.
Also, it’s not related to any e-service properties

Carminepo2 replied:
Leave an empty array

Carminepo2 commented on packages/catalog-process/src/services/catalogService.ts:
```diff
@@ -2237,6 +2290,100 @@ export function catalogServiceBuilder(
+            isConsumerDelegable: false,
+            isClientAccessDelegable: false,
+          },
+          eServiceTemplateReferences: {
+            templateId: template.id,
+            templateVersionId: publishedVersion.id,
+            instanceId: seed.instanceId,
+          },
```
Why not just put all of this directly in the e-service seed?

Alepazz replied:
It was just a choice shared with the team, to be able to reuse the **EServiceSeed** already defined in **catalogApi.yml**.

Carminepo2 replied:
Okkay, fair point, leave it as is then

Alepazz commented on packages/api-clients/open-api/bffApi.yml:
```diff
@@ -19321,6 +19391,18 @@ components:
           $ref: "#/components/schemas/EServiceMode"
         isSignalHubEnabled:
           type: boolean
+    EService:
+      type: object
+      additionalProperties: false
+      required:
+        - id
+        - producerId
+        - name
+        - description
+        - technology
+        - descriptors
+        - riskAnalysis
+        - mode
```
@Carminepo2 
The question is whether the issue is the missing properties or if I wasn't supposed to include the EService at all?

In the first case, I just realized I didn't include the properties because I noticed that by taking them from `catalogApi.yml`, I don't have the `EServiceDescriptor`. Even if I had to include this, I would still be missing some cascading data. How should I proceed?  
If I wasn't supposed to include the `EService`, then I'm not sure what I should have done.

Carminepo2 replied:
It was actually both 😁
In the bff we don't usually return the whole resource on creation.
Let's just return the ids, in this case you can use `CreatedEServiceDescriptor`

Alepazz commented:
@Carminepo2 
I fixed the issue related to **template.audienceDescription** and adjusted the return type for the BFF API (also adding the call on **Bruno**).

Carminepo2 commented on packages/catalog-process/src/routers/EServiceRouter.ts:
```diff
@@ -174,6 +174,32 @@ const eservicesRouter = (
+        } catch (error) {
+          const errorRes = makeApiProblem(
+            error,
+            createEServiceErrorMapper,
```
There should be a custom error mapper that includes the instance creation related error

Carminepo2 replied:
Yes, you'll need to create one specifically for this route

Carminepo2 reviewed (commented):
Great! 🚀
Just a minor change

Carminepo2 commented on packages/catalog-process/src/utilities/errorMappers.ts:
```diff
@@ -27,6 +27,17 @@ export const createEServiceErrorMapper = (
     .with("eServiceDuplicate", () => HTTP_STATUS_CONFLICT)
     .otherwise(() => HTTP_STATUS_INTERNAL_SERVER_ERROR);
 
+export const createEServiceInstanceFromTemplateErrorMapper = (
+  error: ApiError<ErrorCodes>
+): number =>
+  match(error.code)
+    .with("eServiceTemplateNotFound", () => HTTP_STATUS_NOT_FOUND)
+    .with("eServiceTemplateWithoutPublishedVersion", () => HTTP_STATUS_CONFLICT)
+    .with("notValidDescriptor", () => HTTP_STATUS_BAD_REQUEST)
+    .with("interfaceAlreadyExists", () => HTTP_STATUS_CONFLICT)
+    .with("prettyNameDuplicate", () => HTTP_STATUS_CONFLICT)
+    .otherwise(() => HTTP_STATUS_INTERNAL_SERVER_ERROR);
```
Missing errors thrown inside the `innerCreateEService` function 
(originNotCompliant, eServiceDuplicate and inconsistentDailyCalls )

Alepazz replied:
I hate myself, I removed them because I hadn’t noticed they could be thrown.
